src/classes/TriggerHandler.cls

```diff
@@ -0,0 +1,238 @@
+public virtual class TriggerHandler {
+
+  // static map of handlername, times run() was invoked
+  private static Map<String, LoopCount> loopCountMap;
+  private static Set<String> bypassedHandlers;
+
+  // the current context of the trigger, overridable in tests
+  @TestVisible
+  private TriggerContext context;
+
+  // the current context of the trigger, overridable in tests
+  @TestVisible
+  private Boolean isTriggerExecuting;
+
+  // static initialization
+  static {
+    loopCountMap = new Map<String, LoopCount>();
+    bypassedHandlers = new Set<String>();
+  }
+  
+  // constructor
+  public TriggerHandler() {
+    this.setTriggerContext();
+  }
+
+  /***************************************
+   * public instance methods
+   ***************************************/
+
+  // main method that will be called during execution
+  public void run() {
+
+    if(!validateRun()) return;
+
+    addToLoopCount();
+
+    // dispatch to the correct handler method
+    if(this.context == TriggerContext.BEFORE_INSERT) {
+      this.beforeInsert();
+    } else if(this.context == TriggerContext.BEFORE_UPDATE) {
+      this.beforeUpdate();
+    } else if(this.context == TriggerContext.BEFORE_DELETE) {
+      this.beforeDelete();
+    } else if(this.context == TriggerContext.AFTER_INSERT) {
+      this.afterInsert();
+    } else if(this.context == TriggerContext.AFTER_UPDATE) {
+      this.afterUpdate();
+    } else if(this.context == TriggerContext.AFTER_DELETE) {
```

Comment:
Nesting too many `IF` statements affects readability and makes maintenance more difficult. Consider refactoring to simplify the flow of this code block.